Ruler: remote rule evaluation hardening

dannykopping opened the pull request:
**What this PR does / why we need it**:
This PR is part 2 of 2 implementing remote rule evaluation.
See [part 1](https://github.com/grafana/loki/pull/8744) for more context.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
I've added some config to control rule evaluation as well as some metrics.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

JStickler commented on docs/sources/operations/scalability.md:
```diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
+
+## Remote Rule Evaluation
+
+This feature was introduced as part of [`LID-0002`](https://github.com/grafana/loki/pull/8129).
```
I don't think this line is necessary.  
It is also incorrect, the LID is just the design doc and discussion, it doesn't change the code or introduce new features.

dannykopping replied:
You're right, I could've worded it better; I want to include the design doc so folks can read up on the design considerations if they want to.

How's this?

```suggestion
This feature was initial proposed in [`LID-0002`](https://github.com/grafana/loki/pull/8129).
```

JStickler commented on docs/sources/operations/scalability.md:
```diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
+
+This feature was introduced as part of [`LID-0002`](https://github.com/grafana/loki/pull/8129).
+
+### Background & Identification of Problem
```
Your overview text is fine, but I would remove both of the headings:
"Background & Identification of Problem" 
"Recommended Solution"
You're documenting a feature here, not a bug fix.  All of the paragraph text from lines 37 - 51 can sit right under the "Remote rule evaluation" heading.

dannykopping replied:
OK 👍

JStickler commented on docs/sources/operations/scalability.md:
```diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
 
 Configure memory ballast using the ballast_bytes configuration option.
+
+## Remote Rule Evaluation
```
```suggestion
## Remote rule evaluation
```

JStickler commented on docs/sources/operations/scalability.md:
```diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
+
+### Implementation
+
+To enable remote rule evaluation, set the following configuration options:
```
Where do you set this configuration?  In which file?

dannykopping replied:
Loki only supports one configuration file; the other files are for overrides, and it would be clear from context that this belongs in the config file. I can add a link to the config docs page, if that'll make it more clear?

JStickler commented on docs/sources/operations/scalability.md:
```diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
+resource usage of your `ruler`s!). Rules' LogQL queries will be executed against the given `query-frontend` service.
+Requests will be load-balanced across all `query-frontend` IPs if the `dns:///` prefix is used.
+
+Queries that fail to execute are _not_ retried at this time.
```
I think this is worth calling out in a note.
```suggestion
> **Note:** Queries that fail to execute are _not_ retried.
```

dannykopping replied:
Good call 👍

JStickler commented on docs/sources/operations/scalability.md:
````diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
+      address: dns:///<query-frontend-service>:<grpc-port>
+```
+
+The `ruler` component will become simply a gRPC client to the `query-frontend` service (side-note: this will result in much lower
````
```suggestion
When you enable remote rule evaluation, the `ruler` component becomes a gRPC client to the `query-frontend` service.  This may result in lower resource usage of your rulers.
```

dannykopping replied:
Agree with the change except `s/may/will` because the ruler is now doing significantly less work. I'll update it 👍

JStickler commented on docs/sources/operations/scalability.md:
````diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
+```
+
+The `ruler` component will become simply a gRPC client to the `query-frontend` service (side-note: this will result in much lower
+resource usage of your `ruler`s!). Rules' LogQL queries will be executed against the given `query-frontend` service.
````
```suggestion
The LogQL queries coming from the Ruler will be executed against the given `query-frontend` service.
```

JStickler reviewed (commented):
[docs team]

chaudum commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@
 
 ##### Enhancements
 
+* [8785](https://github.com/grafana/loki/pull/8785) **dannykopping**: Ruler: remote rule evaluation hardening.
```
I don't think this needs an extra changelog entry. It's part of the "remote rule evaluation" feature.

dannykopping replied:
That's fair 👍 added it reflexively

chaudum commented on docs/sources/operations/scalability.md:
```diff
@@ -27,3 +27,63 @@ The query scheduler process itself can be started via the `-target=query-schedul
+with gaps or missed alerts. This situation can be detected by alerting on the `cortex_prometheus_rule_group_iterations_missed_total` metric
+when it has a non-zero value.
+
+### Recommended Solution
+A solution to this problem is to externalize rule evaluation from the `ruler` process. The `ruler` embedded query engine
```
Do we have a guideline whether to have a newline between heading and paragraph? I see both variants.

dannykopping replied:
I don't think it matters; the markdown renderer will sort it out - but we could add a linter @jdbaldry?

dannykopping commented on docs/sources/configuration/_index.md:
```diff
@@ -2411,6 +2411,14 @@ ruler_remote_write_sigv4_config:
+[ruler_remote_evaluation_timeout: <duration>]
+
+# Maximum size (in bytes) of the allowable response size from a remote rule
+# evaluation. Set to 0 to disable.
```
This doesn't actually disable remote rule evaluation; it disables the size limitation. I could make that more clear...

How's this?

```suggestion
# evaluation. Set to 0 to allow responses of any size.
```

kavirajk commented on pkg/validation/limits.go:
```diff
@@ -143,6 +143,10 @@ type Limits struct {
 
+	// TODO(dannyk): possible enhancement is to align this with rule group interval
+	RulerRemoteEvaluationTimeout         time.Duration `yaml:"ruler_remote_evaluation_timeout" json:"ruler_remote_evaluation_timeout" doc:"description=Timeout for a remote rule evaluation. Defaults to the value of 'querier.query-timeout'."`
+	RulerRemoteEvaluationMaxResponseSize int64         `yaml:"ruler_remote_evaluation_max_response_size" json:"ruler_remote_evaluation_max_response_size" doc:"description=Maximum size (in bytes) of the allowable response size from a remote rule evaluation. Set to 0 to disable."`
```
Q: I'm failing to find default value for this config (`RulerRemoteEvaluationMaxResponseSize`). Is that intentional? (if not provided, default value will be 0 and meaning ruler remote evaluation will be disabled?)

Looks bit confusing and contradicting with the doc page on how to enable it?

```yaml
ruler:
  evaluation:
    mode: remote
    query_frontend:
      address: dns:///<query-frontend-service>:<grpc-port>
```

This config never mentions anything about setting `RulerRemoteEvaluationMaxResponseSize` when enabling it? Am i missing anything here?

dannykopping replied:
It's an `int64` so its default / zero-value will be `0`.
The wording is misleading here; `Set to 0 to disable` means to disable the limit.

The config is not mentioned in that section above because it's a limit.

chaudum commented on pkg/ruler/evaluator_remote.go:
```diff
@@ -161,36 +227,59 @@ func (q *remoteQuerier) query(ctx context.Context, query string, ts time.Time, l
-	if decoded.Status == "error" {
-		return nil, fmt.Errorf("query response error: %s", decoded.Status)
+	if decoded.Status != loghttp.QueryStatusSuccess {
+		return nil, fmt.Errorf("query response error - status %q: %s", decoded.Status, resp.Body)
```
resp.Body will potentially yield a very long unreadable byte slice. Maybe convert it to `string` first?

dannykopping replied:
It's being interpolated as a string in the `fmt.Errorf`.

kavirajk replied:
Good point. Also I believe `resp.Body` will be of type `io.ReadCloser`, so first reading it with something like `ioutil.ReadAll` and converting to string before printing in logs is the correct way to go I think.

even better to use `io.LimitReader` with some upper limit (just in case for some reason body is huge and to avoid some crazy memory footprints).

kavirajk commented on pkg/ruler/evaluator_remote.go:
```diff
@@ -161,36 +227,59 @@ func (q *remoteQuerier) query(ctx context.Context, query string, ts time.Time, l
+	//   i.e. critical alerts or recording rules producing crucial RemoteEvaluatorMetrics series
 	if resp.Code/100 != 2 {
-		return nil, fmt.Errorf("unexpected response status code %d: %s", resp.Code, string(resp.Body))
+		r.metrics.failedEvals.WithLabelValues("upstream_error", orgID).Inc()
```
nit: Do you think better to put all these reasons (`error`, `upstream_error`, `max_size`, `timeout`) for failed evaluations in one place as package level variables? 

Also these reasons looks so generic? may be tweak a bit? particularly `error`.

dannykopping replied:
> nit: Do you think better to put all these reasons (error, upstream_error, max_size, timeout) for failed evaluations in one place as package level variables?

For what purpose? And I assume you mean constants?

> Also these reasons looks so generic? may be tweak a bit? particularly error.

Correct, we can't have too much cardinality in these labels and also there could be a very large number of possible failure cases. The logs are extensive, so the operator would use the logs to figure out what the actual issue was.

kavirajk replied:
> For what purpose? And I assume you mean constants?

Yep constants. sorry!.

dannykopping replied:
I don't really see the point of having them as package-level constants since their use is localized to this file and there is no duplication.

kavirajk reviewed (approved):
LGTM! few minor nits;

chaudum reviewed (approved):
LGTM

dannykopping commented:
@JStickler I believe I've addressed all your comments, but since we're in different timezones I'm going to merge the PR; if you have any follow-up changes I'll address those in a separate PR as I want to get this code into testing ASAP. Thanks!
